hibernate-core/src/main/java/org/hibernate/tuple/TenantIdGeneration.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.tuple;
+
+import org.hibernate.Session;
+import org.hibernate.annotations.TenantId;
+
+/**
+ * Value generation implementation for {@link TenantId}.
+ *
+ * @author Gavin King
+ */
+public class TenantIdGeneration implements AnnotationValueGeneration<TenantId>, ValueGenerator<Object> {
+
+	@Override
+	public void initialize(TenantId annotation, Class<?> propertyType) {
+	}
+
+	@Override
+	public GenerationTiming getGenerationTiming() {
+		return GenerationTiming.INSERT;
+	}
+
+	@Override
+	public ValueGenerator<?> getValueGenerator() {
+		return this;
+	}
+
+	@Override
+	public Object generateValue(Session session, Object owner) {
+		return session.getTenantIdentifier();
```

Comment:
Does this still allow to set the tenant identifier manually?

Reply:
Looks like it gets overwritten, which is fine, IMO.

Reply:
I've improved the API of `ValueGenerator` to allow this sort of validation.

I now enforce that the tenant id be correct.

Reply:
The improved API looks great, but since this value generator only runs on INSERT, one could still setup cross tenant associations by doing updates, so this is IMO at best a "limitation". Anyway, preventing to accidently set a wrong value is IMO the task of a developer by controlling the write access to the property through proper encapsulation and we shouldn't mess with that. What if a one tenant is allowed to create other (sub-)tenants? It also seems silly to disallow this since one could still do it through an update i.e. I just don't think this kind of validation at this point is useful.

Reply:
Well I was already planning on having `@TenantId` imply `optional=false` and `updateable=false`.

I've implemented that now.

I think it's wrong to call this a "limitation". At most it's a limitation associated with the whole concept of multitenancy based on tenant ids and partitioning. As soon as you have associations between rows with different tenant ids, our filters just aren't going to do the right thing.

> What if a one tenant is allowed to create other (sub-)tenants? 

Well then you would need a whole tree-based tenancy model, not a flat model based on uninterpreted ids and partitioning. And the filters you would need are going to be a *lot* more complicated.

Now, in principle I guess you could now implement something like that on top of the new `AttributeBinder` stuff I just introduced, but you're (quite rightly) on your own.